### ISSUE TITLE
Emulate INSERT .. RETURNING via SQL Server OUTPUT clause

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#4498]
PostgreSQL's 

``` sql
insert into t(v)
values (1), (2), (3)
returning t.*
```

... can be emulated in SQL Server using (in the absence of triggers)

``` sql
insert into t(v)
output INSERTED.*
values (1), (2), (3)
```

... or using (in the presence of triggers)

``` sql
declare @result table (
  v int
);

insert into t(v)
output INSERTED.*
into @result
values (1), (2), (3)

select * from @result
```

Although, triggers can influence the outcome of the `OUTPUT` clause, which makes it useless for the `INSERT .. RETURNING` semantics.

---

See also: #3263

**Comment by kprochazka** [comment]
Hi Lukas, any plans/roadmap for this feature ?

**Comment by lukaseder** [comment]
Thanks for your feedback @kprochazka. This issue hasn't seen much traction by the community, but I'm happy to reconsider this for jOOQ 3.12.

**Comment by kprochazka** [comment]
Thanks @lukaseder. Currently we are using this approach to speed up inserts by using multi-row insert and returning generated ids via output (we do not have any insert triggers). This method proofs to be the most performant as the Microsoft JDBC driver does not support batch insert and generated keys retrieval.

**Comment by lukaseder** [comment]
> This method proofs to be the most performant

I understand

> as the Microsoft JDBC driver does not support batch insert and generated keys retrieval.

Neither does jOOQ :( (see #3327)

**Comment by lukaseder** [comment]
jOOQ currently doesn't maintain any meta information about triggers, so since we cannot really guess the presence of triggers, we'll have to stay on the safe side.

Also, this new feature requires an additional opt out flag in `Settings`, in case a user would prefer to revert to the "old" way of doing this, via `Statement.getGeneratedKeys()`

**Comment by lukaseder** [comment]
This is more subtle than I thought. While this works syntactically:

```sql
declare @result table (
  v int
);

insert into t(v)
output INSERTED.*
into @result
values (1), (2), (3)

select * from @result
```

It will return the `OUTPUT` *prior* to executing any triggers, not after executing them. See: https://docs.microsoft.com/en-us/sql/t-sql/queries/output-clause-transact-sql?view=sql-server-2017

The `RETURNING` clause in other RDBMS, as well as the jOOQ API, expects to produce the output *after* firing triggers, which seems reasonable as triggers are often used to specify default values that are not easy to specify with the DDL `DEFAULT` clause. It is reasonable to return such default values using `RETURNING`.

An emulation using `OUTPUT` (or any other means) would have to take into account this desired functionality. So, this might work better:

```sql
declare @result table (
  id int
);

insert into t(v)
output INSERTED.id
into @result
values (1), (2), (3)

select t where id in (select id from @result)
```

*If* there is an identity column on the table.

**Comment by lukaseder** [comment]
First implementation draft generates the original suggestion. Further investigation is needed. Ideally (for performance reasons), we should know whether there are any triggers, and if not, generate the original suggestion. In the presence of triggers, we can revert to the alternative.

**Comment by lukaseder** [comment]
We need an additional setting which governs whether trigger values should be fetched also in SQL Server (default is true, this flag is to turn off the feature in case it fails or causes performance issues).

When the flag is turned on, we generate:

- The previously mentioned SQL code, if
  - we return the only the primary key 
  - we don't have a primary key
  - the new `Settings.fetchTriggerValuesAfterSQLServerOutput` is `false`
- The below code otherwise:

```sql
-- Declaring the result table for the output clause
declare @result table (
  [id] int,
  [id_generated] int
);

-- Inserting values into a table containing identities (id_generated) and triggers (generating the id value)
insert into [test].[dbo].[t_triggers] ([counter])
output 
  [inserted].[id], 
  [inserted].[id_generated]
into @result
values (0);

-- Fixing the @result table using the previously output primary key value [id_generated]
merge into @result [t]
using (
  (
    select *
    from [test].[dbo].[t_triggers]
  )
) [s]
on [t].[id_generated] = [s].[id_generated]
when matched then update set
  [t].[id] = [s].[id];

-- Fetching back only the columns from the returning clause, 
-- not the possibly augmented columns containing also the primary key
select [@result].[id]
from @result;
```

**Comment by lukaseder** [comment]
This emulation now also works for multi column insert returning statements, e.g.

```java
create().insertInto(TTriggers(), TTriggers_COUNTER())
        .values(-1)
        .values(-2)
        .values(-3)
        .returning()
        .fetch();
```

yields this SQL

```sql
declare @result table (
  [id_generated] int,
  [id] int,
  [counter] int
);
insert into [test].[dbo].[t_triggers] ([counter])
output 
  [inserted].[id_generated], 
  [inserted].[id], 
  [inserted].[counter]
into @result
values (-1), (-2), (-3);
merge into @result [t]
using (
  (
    select *
    from [test].[dbo].[t_triggers]
  )
) [s]
on [t].[id_generated] = [s].[id_generated]
when matched then update set
  [t].[id] = [s].[id], 
  [t].[counter] = [s].[counter];
select 
  [@result].[id_generated], 
  [@result].[id], 
  [@result].[counter]
from @result;
```

**Comment by lukaseder** [comment]
This feature also enables `UPDATE .. RETURNING` and `DELETE .. RETURNING` emulation in SQL Server, using `OUTPUT`: #8920

**Comment by lukaseder** [comment]
Support for returning expressions will be implemented in a separate issue: #8917

**Comment by lukaseder** [comment]
Added support also for `INSERT .. ON DUPLICATE KEY UPDATE .. RETURNING`.

**Input**:

```java
create().insertInto(TAuthor(), TAuthor_ID(), TAuthor_LAST_NAME())
        .values(3, "Rose")
        .onDuplicateKeyUpdate()
        .set(TAuthor_LAST_NAME(), "Christie")
        .returning()
        .fetchOne();
```

**Output**:

```sql
declare @result table (
  [ID] int,
  [FIRST_NAME] varchar(50),
  [LAST_NAME] varchar(50),
  [DATE_OF_BIRTH] date,
  [YEAR_OF_BIRTH] int,
  [ADDRESS] varchar(50)
);
merge into [test].[dbo].[t_author]
using (
  select 1 [one]
) t
on (
  [test].[dbo].[t_author].[ID] = 3
  or [test].[dbo].[t_author].[LAST_NAME] = 'Rose'
)
when matched then update set
  [test].[dbo].[t_author].[LAST_NAME] = 'Christie'
when not matched then insert (
  [ID], 
  [LAST_NAME]
)
values (
  3, 
  'Rose'
)
output 
  [inserted].[ID], 
  [inserted].[FIRST_NAME], 
  [inserted].[LAST_NAME], 
  [inserted].[DATE_OF_BIRTH], 
  [inserted].[YEAR_OF_BIRTH], 
  [inserted].[ADDRESS]
into @result;
merge into @result [r]
using (
  (
    select *
    from [test].[dbo].[t_author]
  )
) [s]
on [r].[ID] = [s].[ID]
when matched then update set
  [r].[FIRST_NAME] = [s].[FIRST_NAME], 
  [r].[LAST_NAME] = [s].[LAST_NAME], 
  [r].[DATE_OF_BIRTH] = [s].[DATE_OF_BIRTH], 
  [r].[YEAR_OF_BIRTH] = [s].[YEAR_OF_BIRTH], 
  [r].[ADDRESS] = [s].[ADDRESS];
select 
  [r].[ID], 
  [r].[FIRST_NAME], 
  [r].[LAST_NAME], 
  [r].[DATE_OF_BIRTH], 
  [r].[YEAR_OF_BIRTH], 
  [r].[ADDRESS]
from @result [r];
```
